### DISCUSSION TITLE
event DataReceived run after get (enter) char at end

### DISCUSSION THREAD

**Discussion by mzelecom** [jchristn/SuperSimpleTcp#92] [General]
hi.
how can i config server to raised DataReceived event when data has (enter) at the end of strings data?
my client is andrid and data change to base64 string .
thanks

**Comment by jchristn** [comment]
Hi @mzelecom you can't.  SimpleTcp will simply read from the socket and fire the event with whatever it gets.  There is no message framing.

You are welcome to modify the DataReceiver method in the client/server to examine the data it's read and fire the event when it suits your needs.

**Comment by mvelayati** [comment]
@jchristn 
Hi Joel,
I know that you envisioned SuperSimpleTCP to stay super simple and not have any type of message framing but it is really common these days to have a "termination" character at the end of transactions when communicating through TCP. 
Would you be opposed to implementing an optional "termination character" feature in SuperSimpleTCP?
I feel like a lot of users would find that useful.

best,
Mo

**Reply by jchristn** [reply]
Hi @mzelecom thanks for your note!  One of the main reasons I chose not to add any kind of framing in SuperSimpleTcp is because WatsonTcp has it.  It provides a benefit in that what you send is what is received on the other end, and I agree that having some kind of termination character would be a simple-enough hint to the other side on when to stop reading and consider what it has received thus far to be a complete message.  

However, I wanted to leave that decision to the developer.  There are a few reasons why:

1) If I pick it, it will likely be a character that someone will need to send when sending binary data through the library
2) If we set it to a well-known (or user-defined) sequence of bytes, then every array read from the network has to be evaluated, or worse, data continues to be buffered until that sequence is encountered, which leads to 3
3) In the event the sequence is received mid-stream, the library then has to take ownership of separating the array (the part before the separator ends the current buffer, and the remainder that has already read, after the separator, starts a new buffer)

Net-net, it introduces a lot of complexity.  For such situations where you need to ensure that what you send is what is received, I'd personally rather go all-in on framing (i.e. WatsonTcp) or a message queue.

Cheers, Joel

**Reply by jchristn** [reply]
If your use case is to fire ```DataReceived``` after each ```\r\n``` is read, it would be easy enough to modify ```DataReceiver``` to do this.  The only major change is maintaining a ```MemoryStream``` and writing data read into it.  After each read, you'd need to convert the stream to a string and check for the existence of the separator.  Everything before the separator gets sent in the ```DataReceived``` event, and anything remaining goes back into the stream.  

If you want help doing this, I'm happy to assist.

**Reply by mvelayati** [reply]
Yeah I'd be happy to help with that!
I'll create a pull request!
